authorization/views.py

```diff
@@ -65,9 +65,9 @@ def handle_no_permission(self):
         # on login and error pages
         error_message(request, message)
 
-        # For authenticated users, raise exception.
+        # For authenticated users, render the exception template.
         if request.user.is_authenticated:
-            raise PermissionDenied(message)
+            return render(template_name="403.html", request=request, status=403)
```

Comment:
The parameters of `render()`: since the `request` parameter is mandatory and not defined as a keyword argument in the `render` function, it would be better coding style to pass the `request` parameter as a positional parameter. In addition, `template_name` is a mandatory positional parameter too.

Note also that the default 403 template name could be read from a constant:
`ERROR_403_TEMPLATE_NAME` (its value is "403.html")
(the constant is defined in django/views/defaults.py)

```python
return render(request, ERROR_403_TEMPLATE_NAME, status=403)
```

https://docs.djangoproject.com/en/3.2/topics/http/shortcuts/#render

I was thinking carefully about why `raise PermissionDenied` does not work here. It causes the HTTP response to use the header `X-Frame-Options DENY`. The header is added even though the `LtiExerciseView` uses the `xframe_options_exempt` decorator. However, Django handles the exception `PermissionDenied` through another view, by default `django.views.defaults.permission_denied`. Since it is another view, the original decorators from the first view do not apply. When this `AuthenticationMixin` class returns a HTTP response directly through `render()`, the code stays inside the original `LtiExerciseView` (that inherits `AuthenticationMixin`) and its decorators apply. The decorator prevents the HTTP response header `X-Frame-Options DENY`.